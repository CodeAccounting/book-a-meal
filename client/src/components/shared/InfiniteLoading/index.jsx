import React, { Component } from 'react';
import PropTypes from 'prop-types';
import InfiniteScroll from 'react-infinite-scroll-component';
import { MiniPreloader } from '../Preloader';

/**
 * @exports
 * @function InfiniteLoading
 * @returns {JSX} InfiniteLoading
 */
class InfiniteLoading extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.any).isRequired,
    limit: PropTypes.number.isRequired
  };

  /**
   * @static
   * @memberof InfiniteLoading
   * @param {object} props
   * @param {object} state
   * @return {object} new state
   */
  static getDerivedStateFromProps(props, state) {
    return { items: props.items.slice(0, state.currentPage * state.limit) };
  }

  /**
   * @memberof InfiniteLoading
   * @constructor
   * @param {object} props
   * @returns {void}
   */
  constructor(props) {
    super(props);

    this.state = {
      items: this.props.items.slice(0, this.props.limit),
      hasMore: this.props.items.length >= this.props.limit,
      currentPage: 1,
      limit: this.props.limit
    };
  }

  /**
   * @memberof InfiniteLoading
   * @returns {void}
   */
  loadMore = () => {
    if (this.state.items.length >= this.props.items.length) {
      this.setState({ hasMore: false });
      return;
    }

    this.setState(prevState => ({
      currentPage: prevState.currentPage + 1
    }), this.loadItems);
  }

  /**
   * @memberof InfiniteLoading
   * @returns {void}
   */
  loadItems = () => {
    const lastIndex = (this.state.currentPage * this.state.limit) - 1;

    this.setState({
      items: this.props.items.slice(0, lastIndex + 1)
    });
  }

  /**
   * @memberof InifinteScroll
   * @returns {JSX} InfiniteLoading
   */
  render() {
    return (
      <InfiniteScroll
        dataLength={this.state.items.length}
        next={this.loadMore}
        hasMore={this.state.hasMore}
        loader={<div className="text-center"><MiniPreloader /></div>}
      >
        <div className="scroller">
          {this.state.items}
        </div>
      </InfiniteScroll>
    );
  }
}

export default InfiniteLoading;
