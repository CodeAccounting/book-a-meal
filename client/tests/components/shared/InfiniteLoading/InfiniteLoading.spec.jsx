import React from 'react';
import InfiniteLoading from '../../../../src/components/shared/InfiniteLoading';

describe('InfiniteLoading', () => {
  const items = Array.from({ length: 10 }, (v, i) => <p key={i}>Heya!</p>);
  it('renders correctly', () => {
    const wrapper = mount(<InfiniteLoading items={items} limit={3} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('updates current page when there are more elements to load', () => {
    const wrapper = mount(<InfiniteLoading items={items} limit={3} />);

    expect(wrapper.state().currentPage).toEqual(1);

    wrapper.instance().loadMore();

    expect(wrapper.state().currentPage).toEqual(2);
  });

  it('does not update current page when there are no more elements to load', () => {
    const wrapper = mount(<InfiniteLoading items={items} limit={10} />);

    expect(wrapper.state().currentPage).toEqual(1);

    wrapper.instance().loadMore();

    expect(wrapper.state().currentPage).toEqual(1);
    expect(wrapper.state().hasMore).toEqual(false);
  });
});
