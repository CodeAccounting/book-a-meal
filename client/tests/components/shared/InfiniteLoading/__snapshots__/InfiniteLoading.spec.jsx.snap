// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteLoading renders correctly 1`] = `
<InfiniteLoading
  items={
    Array [
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
      <p>
        Heya!
      </p>,
    ]
  }
  limit={3}
>
  <InfiniteScroll
    dataLength={3}
    disableBrowserPullToRefresh={true}
    hasMore={true}
    loader={
      <div
        className="text-center"
      >
        <MiniPreloader />
      </div>
    }
    next={[Function]}
    pullDownToRefreshContent={
      <h3>
        Pull down to refresh
      </h3>
    }
    pullDownToRefreshThreshold={100}
    releaseToRefreshContent={
      <h3>
        Release to refresh
      </h3>
    }
  >
    <div
      style={Object {}}
    >
      <div
        className="infinite-scroll-component"
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="scroller"
        >
          <p
            key="0"
          >
            Heya!
          </p>
          <p
            key="1"
          >
            Heya!
          </p>
          <p
            key="2"
          >
            Heya!
          </p>
        </div>
        <div
          className="text-center"
        >
          <MiniPreloader>
            <Loader
              active={true}
              type="ball-beat"
            >
              <div
                className="loader loader-undefined loader-active"
              >
                <div
                  className="loader-inner ball-beat"
                >
                  <div
                    key="0"
                    style={Object {}}
                  />
                  <div
                    key="1"
                    style={Object {}}
                  />
                  <div
                    key="2"
                    style={Object {}}
                  />
                </div>
              </div>
            </Loader>
          </MiniPreloader>
        </div>
      </div>
    </div>
  </InfiniteScroll>
</InfiniteLoading>
`;
